trieste/models/keras/models.py

```diff
@@ -367,4 +371,321 @@ def __setstate__(self, state: dict[str, Any]) -> None:
+            "The main predictor is meant to use a :class:`TrainableProbabilisticModel`."
+        )
+        assert issubclass(type(self._model["e_model"]), tf.keras.Model), (
+            "The epistemic predictor is meant to use :class:`~tf.keras.Model`."
+        )
+        return self._model["f_model"], self._model["e_model"]
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        """
+        Return ``num_samples`` samples at ``query_points``. We use the mixture approximation in
+        :meth:`predict` for ``query_points`` and sample ``num_samples`` times from a Gaussian
+        distribution given by the predicted mean and variance.
+
+        :param query_points: The points at which to sample, with shape [..., N, D].
+        :param num_samples: The number of samples at each point.
+        :return: The samples. For a predictive distribution with event shape E, this has shape
+            [..., S, N] + E, where S is the number of samples.
+        """
+
+        predicted_means, predicted_vars = self.predict(query_points)
+        normal = tfp.distributions.Normal(predicted_means, tf.sqrt(predicted_vars))
+        samples = normal.sample(num_samples)
+
+        return samples  # [num_samples, len(query_points), 1]
+
+    def update(self, dataset: Dataset) -> None:
+        """
+        Neural networks are parametric models and do not need to update data.
+        `TrainableProbabilisticModel` interface, however, requires an update method, so
+        here we simply pass the execution.
+        """
+        pass
+
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize the underlying Direct Epistemic Uncertainty Prediction model with the 
+        specified ``dataset``.
+
+        Optimization is performed by using the Keras `fit` method, rather than applying the
+        optimizer and using the batches supplied with the optimizer wrapper. User can pass
+        arguments to the `fit` method through ``minimize_args`` argument in the optimizer wrapper.
+        These default to using 100 epochs, batch size 32, and verbose 0. See
+        https://keras.io/api/models/model_training_apis/#fit-method for a list of possible
+        arguments.
+
+        Note that optimization does not return the result, instead optimization results are
+        stored in a history attribute of the model object. The algorithm iterates
+        over two copies of the dataset's observations to account for the
+        stationarizing features and the two targets: the target outcome for the main
+        predictor and the squared loss for the auxiliary predictor. The procedure follows
+        the main proposed algorithm in <cite data-cite="jain2022"/>.
+
+        :param dataset: The data with which to optimize the model.
+        """     
+        # optional init buffer
+        if self._data_u is None:
+            self.density_estimator = KernelDensityEstimator(kernel="gaussian")
```

Comment:
Should kernel type be a hyperparameter? if not why even bother with any argument at all?

Reply:
Same with DeepEnsembles, to facilitate / remind me that this should eventually be a hyperparameter. But indeed no HP at the moment, so I'll simplify it.